File: app/cqrs/core/bus/EventBus.java
package cqrs.core.bus;

import cqrs.core.Event;
import cqrs.core.EventPublisher;
import cqrs.core.Handler;
import cqrs.core.Message;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class EventBus implements EventPublisher {
    private final HashMap<Class<?>, List<Handler>> routes = new HashMap<Class<?>,List<Handler>>();

    public <T extends Message> void registerHandler(Class<T> type, Handler<T> handler) {
        List<Handler> handlers = routes.get(type);

        if (handlers==null) {
            handlers = new ArrayList<Handler>();
            handlers.add(handler);
            routes.put(type, handlers);
        } else {
            handlers.add(handler);
        }
    }

    public <T extends Event> void publish(T event) {
        List<Handler> handlers = routes.get(event.getClass());

        if (handlers == null) return;

        for(Handler handler : handlers) {
            handler.handle(event);
        }
    }
}

File: app/cqrs/core/bus/EventBus.scala
package cqrs.core.bus

import cqrs.core.{Event, EventPublisher, Handler}

class EventBus extends AbstractBus with EventPublisher {
  def publish[T <: Event](event: T): Unit = {
    routes.get(event.getClass()) match {
      case Some(handlers) =>
        val hs = handlers.asInstanceOf[List[Handler[T]]]
        hs.foreach { handler =>
          handler.handle(event)
        }
      case None =>
    }
  }
}
